app/javascript/src/components/Editor.tsx

```diff
@@ -363,6 +375,148 @@ class Editor extends React.Component<EditorProps, EditorState> {
     }
   }
 
+  private onCopy = () => {
+    const { model } = this
+
+    const selectedNodes = model.getSelectedItems().filter(item => {
+      return item instanceof DefaultNodeModel
+    }) as DefaultNodeModel[]
+    const selectedLinks = this.model.getSelectedItems().filter(item => {
+      return item instanceof DefaultLinkModel
+    }) as DefaultLinkModel[]
+
+    this.copiedLinks = selectedLinks
+    this.copiedNodes = selectedNodes
+    this.pastedNodes = []
+    this.pastedLinks = []
+  }
+
+  private onPaste = () => {
+    const { copiedNodes, model, pastedNodes, pastedLinks, copiedLinks } = this
+
+    model.clearSelection()
+
+    for (let node of copiedNodes) {
+      let copiedNode = this.createCopiedNode(
+        node,
+        node.x + offset,
+        node.y + offset
+      )
+      if (copiedNode) {
+        pastedNodes.push(copiedNode)
+      }
+    }
+    for (let link of copiedLinks) {
+      let copiedLink = this.createCopiedLink(link)
+      if (copiedLink) {
+        pastedLinks.push(copiedLink)
+      }
+    }
+
+    for (let node of pastedNodes) {
+      model.addNode(node)
+    }
+    for (let link of pastedLinks) {
```

Comment:
Could this be `pastedLinks.forEach(model.addLink)`? Maybe I just have a superstition about this, but `array.[functional method]()` tends to be nicer/clearer than for loops in JS.